apache/incubator-kie-tools
# KOGITO-7066: SWF - Expose Stunner API through multiplying architecture

See [KOGITO-7066](https://issues.redhat.com/browse/KOGITO-7066)

**How to test the API from the UI**

Notice the API is implemented at _combined editor_ level, so it is accessible to most of the channels, except VSCode extension. Here will share some instructions about how to test it by using the _web-tools_ channel.

1. Download and extract the `build_artifacts.zip` , from this PR (_Checks_ tab)
2. Run a web server in the root folder: `build_artifacts/packages/serverless-logic-web-tools/dist`. _I personally use the Live Server extension in VSCode, I just open the dist/ folder and run the server._
4. Open `index.html` in your browser (located at root `build_artifacts/packages/serverless-logic-web-tools/dist/index.html`)
5. Once having the web-tools UI available, just create or open a workflow
6. Open the chrome _Console_ tab, from the web tools
7. TIP: Some errors are displayed, in order not to make noise, please filter our errors  from the console, Unselect the `errors` item here:
![stunner-envelope-api_filterout_errors](https://github.com/kiegroup/kie-tools/assets/4602417/d2a79e44-eb84-43c4-be14-548ffdddecbb)
8. In the console, choose the right context by selecting the `serverless-workflow-combined-editor  iframe `.  See:
![stunner-envelope-api_frame_selection](https://github.com/kiegroup/kie-tools/assets/4602417/35fb7ae9-3db3-4dc4-b7a2-b72f883995a1)



**Examples**
Notice the API is documented at https://github.com/kiegroup/kie-tools/pull/1589/files#diff-292fd900d04b40fbd39f26fefa506b17357917982fce264719152ed0ee62d913R23
From this point, you can use the API in the console, here are some examples:
**a) Obtain the selected node UIUID:**
a.1) Select a node from the UI, by clicking on any state
a.2) `editor.session.getSelectedElementUUID().then(uuid => console.log(uuid))`

**b) Change colors for an state**
b.1) Ensure nothing is selected on the UI
b.2) `editor.canvas.setBackgroundColor('_0757D1D8-C613-4B1A-955E-300BB2818D92', 'grey').then(done => {})`
b.3) `editor.canvas.setBorderColor('_0757D1D8-C613-4B1A-955E-300BB2818D92', 'black').then(done => {})`
b.4) `editor.canvas.draw().then(done => {})`
b.5) See results:
![stunner-envelope-api_canvas_example](https://github.com/kiegroup/kie-tools/assets/4602417/4956e5f8-59f8-40bd-8314-0635c8164754)
